From 99f48528dc386c73e1a5e93182c6a32a1583a195 Mon Sep 17 00:00:00 2001
From: Ofir Fedida <ofedida@marvell.com>
Date: Wed, 13 Jan 2016 13:52:05 +0200
Subject: [PATCH 0686/1240] a3700: ddr: Add DLL tuning to the initialization
 sequence

- Create a DDR driver for a3700 platform based on A8K MC6
  driver code
- Implement DLL tuning procdure inside the new DDR driver.
  Call to DLL tuning procedure in SPL context.
- Current code for DDR PHY and MAC initialization is not
  called, it has to be reviewed and activated later
- Add definitions for DDR MAC and PHY base addresses
  according to MC6 template to a3700 DTSI common file
- Add definition for DDR MAC with single CS (512MB)
  to a3700 DB DTS

Change-Id: I5be818c2e83f6a09f9bde42bc68682fa158e54c0
Signed-off-by: Ofir Fedida <ofedida@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/26791
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
---
 arch/arm/dts/armada-lp-db.dts         |    9 +-
 arch/arm/dts/armada-lp.dtsi           |   11 +
 arch/arm/include/asm/arch-mvebu/ddr.h |    3 +-
 drivers/ddr/ddr_mckinley6_arlp.c      | 1188 ++++++---------------------------
 drivers/ddr/mvebu_ddr.c               |    5 +-
 5 files changed, 213 insertions(+), 1003 deletions(-)

diff --git a/arch/arm/dts/armada-lp-db.dts b/arch/arm/dts/armada-lp-db.dts
index c82270d..c0d0413 100644
--- a/arch/arm/dts/armada-lp-db.dts
+++ b/arch/arm/dts/armada-lp-db.dts
@@ -1,6 +1,6 @@
 
 /dts-v1/;
-
+#include <dt-bindings/ddr/ddr_data.h>
 #include "armada-lp.dtsi"
 
 / {
@@ -13,6 +13,13 @@
 		};
 
 		internal-regs {
+			ddr-mac {
+				cs_count = <1>;
+				bus_width = <DDR_BUS_WIDTH_16>;
+				cs0 {
+					size = <0x00000000 0x20000000>;
+				};
+			};
 			serial: serial@12000 {
 				status = "okay";
 			};
diff --git a/arch/arm/dts/armada-lp.dtsi b/arch/arm/dts/armada-lp.dtsi
index 605c9f4..d2bb9da 100644
--- a/arch/arm/dts/armada-lp.dtsi
+++ b/arch/arm/dts/armada-lp.dtsi
@@ -130,6 +130,17 @@
 				reg = <0x58000 0x450>;
 				status = "disabled";
 			};
+			ddr-mac {
+				compatible = "marvell,mvebu-ddr-mac";
+				reg = <0x0000 0x1000>;
+				status = "okay";
+			};
+
+			ddr-phy {
+				compatible = "marvell,mvebu-ddr-phy";
+				reg = <0x1000 0x1000>;
+				status = "okay";
+			};
 		};
 	};
 };
diff --git a/arch/arm/include/asm/arch-mvebu/ddr.h b/arch/arm/include/asm/arch-mvebu/ddr.h
index 80767ec..fbc5270 100644
--- a/arch/arm/include/asm/arch-mvebu/ddr.h
+++ b/arch/arm/include/asm/arch-mvebu/ddr.h
@@ -35,9 +35,8 @@ struct mvebu_dram_config {
 	u32 bus_width;
 	u32 size_mb;
 };
-
 void mvebu_dram_init(const void *blob);
 void mvebu_dram_mac_init(struct mvebu_dram_config *dram_config);
 void mvebu_dram_phy_init(struct mvebu_dram_config *dram_config);
-
+void mvebu_dram_dll_tune(struct mvebu_dram_config *dram_config);
 #endif /* _MVEBU_DDR_H_ */
diff --git a/drivers/ddr/ddr_mckinley6_arlp.c b/drivers/ddr/ddr_mckinley6_arlp.c
index d6eb521..f1b2807 100644
--- a/drivers/ddr/ddr_mckinley6_arlp.c
+++ b/drivers/ddr/ddr_mckinley6_arlp.c
@@ -40,11 +40,20 @@
 #define MMAP_AREA_LEN_MASK	(0x1F << 16)
 #define MMAP_AREA_LEN(x)	((x) << 16)
 
-#define mck6_writel(v, c)					\
-	do {							\
-		debug("0x%p - 0x08%x\n", c, v);			\
-		writel(v, c);					\
-	} while (0)
+/* DLL Tune definitions */
+#define DLL_PHSEL_START			0x00
+#define DLL_PHSEL_END			0x3F
+#define DLL_PHSEL_STEP			0x1
+#define MC6_CH0_PHY_CONTROL_8		(0x1C)
+#define MC6_CH0_PHY_CONTROL_9		(0x20)
+#define MC6_CH0_PHY_DLL_CONTROL_B0	(0x50)
+#define SDRAM_DIRECT_START		(0x6000000)
+
+#define mck6_writel(v, c)		\
+do {							\
+	debug("0x%p - 0x08%x\n", c, v);	\
+	writel(v, c);					\
+} while (0)
 
 enum mvebu_mck_freq_support {
 	FREQ_650_HZ = 0,
@@ -57,1004 +66,10 @@ struct mvebu_mckinley_config {
 };
 
 struct mvebu_mckinley_config mckinley_mac_config[] = {
-/*  reg_offset  650_HZ */
-	{ 0x000, { 0x06501140} },
-	{ 0x004, { 0x000f9fff} },
-	{ 0x008, { 0x00000001} },
-	{ 0x00c, { 0x00000000} },
-	{ 0x010, { 0x00000000} },
-	{ 0x014, { 0x00000000} },
-	{ 0x018, { 0x00000000} },
-	{ 0x01c, { 0x00000000} },
-	{ 0x020, { 0x00000000} },
-	{ 0x024, { 0x00000000} },
-	{ 0x028, { 0x00000000} },
-	{ 0x02c, { 0x00000000} },
-	{ 0x030, { 0x00000000} },
-	{ 0x034, { 0x00000000} },
-	{ 0x038, { 0x00000000} },
-	{ 0x03c, { 0x00000000} },
-	{ 0x040, { 0x22222022} },
-	{ 0x044, { 0x00030300} },
-	{ 0x048, { 0x00000001} },
-	{ 0x04c, { 0x00000000} },
-	{ 0x050, { 0x00000000} },
-	{ 0x054, { 0x00000480} },
-	{ 0x058, { 0x00003ff5} },
-	{ 0x05c, { 0x00000000} },
-	{ 0x060, { 0x00000000} },
-	{ 0x064, { 0x0000000e} },
-	{ 0x068, { 0x00000000} },
-	{ 0x06c, { 0x00000000} },
-	{ 0x070, { 0x00000000} },
-	{ 0x074, { 0x00000000} },
-	{ 0x078, { 0x00000000} },
-	{ 0x07c, { 0x00000000} },
-	{ 0x080, { 0x00000000} },
-	{ 0x084, { 0x00000000} },
-	{ 0x088, { 0x00000000} },
-	{ 0x08c, { 0x00000000} },
-	{ 0x090, { 0x00000000} },
-	{ 0x094, { 0x00000000} },
-	{ 0x098, { 0x0040000e} },
-	{ 0x09c, { 0x00000000} },
-	{ 0x0a0, { 0x00000000} },
-	{ 0x0a4, { 0x00000000} },
-	{ 0x0a8, { 0x00000000} },
-	{ 0x0ac, { 0x00000000} },
-	{ 0x0b0, { 0x00000000} },
-	{ 0x0b4, { 0x00000000} },
-	{ 0x0b8, { 0x00000000} },
-	{ 0x0bc, { 0x00000000} },
-	{ 0x0c0, { 0x00000000} },
-	{ 0x0c4, { 0x00000000} },
-	{ 0x0c8, { 0x00000000} },
-	{ 0x0cc, { 0x00000000} },
-	{ 0x0d0, { 0x00000000} },
-	{ 0x0d4, { 0x00000000} },
-	{ 0x0d8, { 0x00000000} },
-	{ 0x0dc, { 0x00000000} },
-	{ 0x0e0, { 0x00000000} },
-	{ 0x0e4, { 0x00000000} },
-	{ 0x0e8, { 0x00000000} },
-	{ 0x0ec, { 0x00000000} },
-	{ 0x0f0, { 0x00000000} },
-	{ 0x0f4, { 0x00000000} },
-	{ 0x0f8, { 0x00000000} },
-	{ 0x0fc, { 0x00000000} },
-	{ 0x100, { 0x00000000} },
-	{ 0x104, { 0x00000000} },
-	{ 0x108, { 0x00000000} },
-	{ 0x10c, { 0x00000000} },
-	{ 0x110, { 0x00000000} },
-	{ 0x114, { 0x00000000} },
-	{ 0x118, { 0x00000000} },
-	{ 0x11c, { 0x00000000} },
-	{ 0x120, { 0x00000000} },
-	{ 0x124, { 0x00000000} },
-	{ 0x128, { 0x00000000} },
-	{ 0x12c, { 0x00000000} },
-	{ 0x130, { 0x00000000} },
-	{ 0x134, { 0x00000000} },
-	{ 0x138, { 0x00000000} },
-	{ 0x13c, { 0x00000000} },
-	{ 0x140, { 0x00000000} },
-	{ 0x144, { 0x00000000} },
-	{ 0x148, { 0x00000000} },
-	{ 0x14c, { 0x00000000} },
-	{ 0x150, { 0x00000000} },
-	{ 0x154, { 0x00000000} },
-	{ 0x158, { 0x00000000} },
-	{ 0x15c, { 0x00000000} },
-	{ 0x160, { 0x00010000} },
-	{ 0x164, { 0x00005014} },
-	{ 0x168, { 0x00000000} },
-	{ 0x16c, { 0x00000000} },
-	{ 0x170, { 0x00000000} },
-	{ 0x174, { 0x00000000} },
-	{ 0x178, { 0x00000000} },
-	{ 0x17c, { 0x00000000} },
-	{ 0x180, { 0x00000000} },
-	{ 0x184, { 0x00000000} },
-	{ 0x188, { 0x00000000} },
-	{ 0x18c, { 0x00000000} },
-	{ 0x190, { 0x00000000} },
-	{ 0x194, { 0x00000000} },
-	{ 0x198, { 0x00000000} },
-	{ 0x19c, { 0x00000000} },
-	{ 0x1a0, { 0x00000000} },
-	{ 0x1a4, { 0x00000000} },
-	{ 0x1a8, { 0x00000000} },
-	{ 0x1ac, { 0x00000000} },
-	{ 0x1b0, { 0x00000000} },
-	{ 0x1b4, { 0x00000000} },
-	{ 0x1b8, { 0x00000000} },
-	{ 0x1bc, { 0x00000000} },
-	{ 0x1c0, { 0x00000000} },
-	{ 0x1c4, { 0x00000000} },
-	{ 0x1c8, { 0x00000000} },
-	{ 0x1cc, { 0x00000000} },
-	{ 0x1d0, { 0x00000000} },
-	{ 0x1d4, { 0x00000000} },
-	{ 0x1d8, { 0x00000000} },
-	{ 0x1dc, { 0x00000000} },
-	{ 0x1e0, { 0x00000000} },
-	{ 0x1e4, { 0x00000000} },
-	{ 0x1e8, { 0x00000000} },
-	{ 0x1ec, { 0x00000000} },
-	{ 0x1f0, { 0x00000000} },
-	{ 0x1f4, { 0x00000000} },
-	{ 0x1f8, { 0x00000000} },
-	{ 0x1fc, { 0x00000000} },
-	{ 0x200, { 0x00100001} },
-	{ 0x204, { 0x00000000} },
-	{ 0x208, { 0x00000000} },
-	{ 0x20c, { 0x00000000} },
-	{ 0x210, { 0x00000000} },
-	{ 0x214, { 0x00000000} },
-	{ 0x218, { 0x00000000} },
-	{ 0x21c, { 0x00000000} },
-	{ 0x220, { 0x11010642} },
-	{ 0x224, { 0x02000332} },
-	{ 0x228, { 0x02000332} },
-	{ 0x22c, { 0x02000332} },
-	{ 0x230, { 0x00000000} },
-	{ 0x234, { 0x00000000} },
-	{ 0x238, { 0x00000000} },
-	{ 0x23c, { 0x00000000} },
-	{ 0x240, { 0x00000000} },
-	{ 0x244, { 0x00000000} },
-	{ 0x248, { 0x00000000} },
-	{ 0x24c, { 0x00000000} },
-	{ 0x250, { 0x00000000} },
-	{ 0x254, { 0x00000000} },
-	{ 0x258, { 0x00000000} },
-	{ 0x25c, { 0x00000000} },
-	{ 0x260, { 0x00000000} },
-	{ 0x264, { 0x00000000} },
-	{ 0x268, { 0x00000000} },
-	{ 0x26c, { 0x00000000} },
-	{ 0x270, { 0x00000000} },
-	{ 0x274, { 0x00000000} },
-	{ 0x278, { 0x00000000} },
-	{ 0x27c, { 0x00000000} },
-	{ 0x280, { 0x00000000} },
-	{ 0x284, { 0x00000000} },
-	{ 0x288, { 0x00000000} },
-	{ 0x28c, { 0x00000000} },
-	{ 0x290, { 0x00000000} },
-	{ 0x294, { 0x00000000} },
-	{ 0x298, { 0x00000000} },
-	{ 0x29c, { 0x00000000} },
-	{ 0x2a0, { 0x00000000} },
-	{ 0x2a4, { 0x00000000} },
-	{ 0x2a8, { 0x00000000} },
-	{ 0x2ac, { 0x00000000} },
-	{ 0x2b0, { 0x00000000} },
-	{ 0x2b4, { 0x00000000} },
-	{ 0x2b8, { 0x00000000} },
-	{ 0x2bc, { 0x00000000} },
-	{ 0x2c0, { 0x0000e000} },
-	{ 0x2c4, { 0x00000020} },
-	{ 0x2c8, { 0x00000000} },
-	{ 0x2cc, { 0x00000000} },
-	{ 0x2d0, { 0x00000000} },
-	{ 0x2d4, { 0x00000000} },
-	{ 0x2d8, { 0x00000000} },
-	{ 0x2dc, { 0x00000000} },
-	{ 0x2e0, { 0x00000000} },
-	{ 0x2e4, { 0x00000000} },
-	{ 0x2e8, { 0x00000000} },
-	{ 0x2ec, { 0x00000000} },
-	{ 0x2f0, { 0x00000000} },
-	{ 0x2f4, { 0x00000000} },
-	{ 0x2f8, { 0x00000000} },
-	{ 0x2fc, { 0x00000000} },
-	{ 0x300, { 0x00000709} },
-	{ 0x304, { 0x00000400} },
-	{ 0x308, { 0x00000000} },
-	{ 0x30c, { 0x00000000} },
-	{ 0x310, { 0x00110000} },
-	{ 0x314, { 0x00000000} },
-	{ 0x318, { 0x00000000} },
-	{ 0x31c, { 0x00000000} },
-	{ 0x320, { 0x00000000} },
-	{ 0x324, { 0x00000000} },
-	{ 0x328, { 0x00000000} },
-	{ 0x32c, { 0x00000000} },
-	{ 0x330, { 0x00000000} },
-	{ 0x334, { 0x00000000} },
-	{ 0x338, { 0x00000000} },
-	{ 0x33c, { 0x00000000} },
-	{ 0x340, { 0x00000001} },
-	{ 0x344, { 0x00000002} },
-	{ 0x348, { 0x00000000} },
-	{ 0x34c, { 0x00000000} },
-	{ 0x350, { 0x00000000} },
-	{ 0x354, { 0x00000000} },
-	{ 0x358, { 0x00000000} },
-	{ 0x35c, { 0x00000000} },
-	{ 0x360, { 0x00000000} },
-	{ 0x364, { 0x00000000} },
-	{ 0x368, { 0x00000000} },
-	{ 0x36c, { 0x00000000} },
-	{ 0x370, { 0x00000000} },
-	{ 0x374, { 0x00000000} },
-	{ 0x378, { 0x00000000} },
-	{ 0x37c, { 0x00000000} },
-	{ 0x380, { 0x000516bc} },
-	{ 0x384, { 0x00020918} },
-	{ 0x388, { 0x00000043} },
-	{ 0x38c, { 0x00000200} },
-	{ 0x390, { 0x00400100} },
-	{ 0x394, { 0x00ea0517} },
-	{ 0x398, { 0x00f10200} },
-	{ 0x39c, { 0x00000707} },
-	{ 0x3a0, { 0x00050511} },
-	{ 0x3a4, { 0x00000002} },
-	{ 0x3a8, { 0x00000c04} },
-	{ 0x3ac, { 0x18200a17} },
-	{ 0x3b0, { 0x0a0b060a} },
-	{ 0x3b4, { 0x04000600} },
-	{ 0x3b8, { 0x00000500} },
-	{ 0x3bc, { 0x00050000} },
-	{ 0x3c0, { 0x00000000} },
-	{ 0x3c4, { 0x00000000} },
-	{ 0x3c8, { 0x00000000} },
-	{ 0x3cc, { 0x00000000} },
-	{ 0x3d0, { 0x00000000} },
-	{ 0x3d4, { 0x0d380a03} },
-	{ 0x3d8, { 0x0000012d} },
-	{ 0x3dc, { 0x00000000} },
-	{ 0x3e0, { 0x00000000} },
-	{ 0x3e4, { 0x00000000} },
-	{ 0x3e8, { 0x00000000} },
-	{ 0x3ec, { 0x00000000} },
-	{ 0x3f0, { 0x00000000} },
-	{ 0x3f4, { 0x00000000} },
-	{ 0x3f8, { 0x00000000} },
-	{ 0x3fc, { 0x00000000} },
-	{ 0x400, { 0x00000000} },
-	{ 0x404, { 0x00000000} },
-	{ 0x408, { 0x00000000} },
-	{ 0x40c, { 0x00000000} },
-	{ 0x410, { 0x00000000} },
-	{ 0x414, { 0x00000000} },
-	{ 0x418, { 0x00000000} },
-	{ 0x41c, { 0x00000000} },
-	{ 0x420, { 0x00000000} },
-	{ 0x424, { 0x00000000} },
-	{ 0x428, { 0x00000000} },
-	{ 0x42c, { 0x00000000} },
-	{ 0x430, { 0x00000000} },
-	{ 0x434, { 0x00000000} },
-	{ 0x438, { 0x00000000} },
-	{ 0x43c, { 0x00000000} },
-	{ 0x440, { 0x00000000} },
-	{ 0x444, { 0x00000000} },
-	{ 0x448, { 0x00000000} },
-	{ 0x44c, { 0x00000000} },
-	{ 0x450, { 0x00000000} },
-	{ 0x454, { 0x00000000} },
-	{ 0x458, { 0x00000000} },
-	{ 0x45c, { 0x00000000} },
-	{ 0x460, { 0x00000000} },
-	{ 0x464, { 0x00000000} },
-	{ 0x468, { 0x00000000} },
-	{ 0x46c, { 0x00000000} },
-	{ 0x470, { 0x00000000} },
-	{ 0x474, { 0x00000000} },
-	{ 0x478, { 0x00000000} },
-	{ 0x47c, { 0x00000000} },
-	{ 0x480, { 0x00000000} },
-	{ 0x484, { 0x00000000} },
-	{ 0x488, { 0x00000000} },
-	{ 0x48c, { 0x00000000} },
-	{ 0x490, { 0x00000000} },
-	{ 0x494, { 0x00000000} },
-	{ 0x498, { 0x00000000} },
-	{ 0x49c, { 0x00000000} },
-	{ 0x4a0, { 0x00000000} },
-	{ 0x4a4, { 0x00000000} },
-	{ 0x4a8, { 0x00000000} },
-	{ 0x4ac, { 0x00000000} },
-	{ 0x4b0, { 0x00000000} },
-	{ 0x4b4, { 0x00000000} },
-	{ 0x4b8, { 0x00000000} },
-	{ 0x4bc, { 0x00000000} },
-	{ 0x4c0, { 0x00000000} },
-	{ 0x4c4, { 0x00000000} },
-	{ 0x4c8, { 0x00000000} },
-	{ 0x4cc, { 0x00000000} },
-	{ 0x4d0, { 0x00000000} },
-	{ 0x4d4, { 0x00000000} },
-	{ 0x4d8, { 0x00000000} },
-	{ 0x4dc, { 0x00000000} },
-	{ 0x4e0, { 0x00000000} },
-	{ 0x4e4, { 0x00000000} },
-	{ 0x4e8, { 0x00000000} },
-	{ 0x4ec, { 0x00000000} },
-	{ 0x4f0, { 0x00000000} },
-	{ 0x4f4, { 0x00000000} },
-	{ 0x4f8, { 0x00000000} },
-	{ 0x4fc, { 0x00000000} },
-	{ 0x500, { 0x00000000} },
-	{ 0x504, { 0x00000000} },
-	{ 0x508, { 0x00000000} },
-	{ 0x50c, { 0x00000000} },
-	{ 0x510, { 0x00000000} },
-	{ 0x514, { 0x00000000} },
-	{ 0x518, { 0x00000000} },
-	{ 0x51c, { 0x00000000} },
-	{ 0x520, { 0x00000000} },
-	{ 0x524, { 0x00000000} },
-	{ 0x528, { 0x00000000} },
-	{ 0x52c, { 0x00000000} },
-	{ 0x530, { 0x00000000} },
-	{ 0x534, { 0x00000000} },
-	{ 0x538, { 0x00000000} },
-	{ 0x53c, { 0x00000000} },
-	{ 0x540, { 0x00000000} },
-	{ 0x544, { 0x00000000} },
-	{ 0x548, { 0x00000000} },
-	{ 0x54c, { 0x00000000} },
-	{ 0x550, { 0x00000000} },
-	{ 0x554, { 0x00000000} },
-	{ 0x558, { 0x00000000} },
-	{ 0x55c, { 0x00000000} },
-	{ 0x560, { 0x00000000} },
-	{ 0x564, { 0x00000000} },
-	{ 0x568, { 0x00000000} },
-	{ 0x56c, { 0x00000000} },
-	{ 0x570, { 0x00000000} },
-	{ 0x574, { 0x00000000} },
-	{ 0x578, { 0x00000000} },
-	{ 0x57c, { 0x00000000} },
-	{ 0x580, { 0x00000000} },
-	{ 0x584, { 0x00000000} },
-	{ 0x588, { 0x00000000} },
-	{ 0x58c, { 0x00000000} },
-	{ 0x590, { 0x00000000} },
-	{ 0x594, { 0x00000000} },
-	{ 0x598, { 0x00000000} },
-	{ 0x59c, { 0x00000000} },
-	{ 0x5a0, { 0x00000000} },
-	{ 0x5a4, { 0x00000000} },
-	{ 0x5a8, { 0x00000000} },
-	{ 0x5ac, { 0x00000000} },
-	{ 0x5b0, { 0x00000000} },
-	{ 0x5b4, { 0x00000000} },
-	{ 0x5b8, { 0x00000000} },
-	{ 0x5bc, { 0x00000000} },
-	{ 0x5c0, { 0x00000000} },
-	{ 0x5c4, { 0x00000000} },
-	{ 0x5c8, { 0x00000000} },
-	{ 0x5cc, { 0x00000000} },
-	{ 0x5d0, { 0x00000000} },
-	{ 0x5d4, { 0x00000000} },
-	{ 0x5d8, { 0x00000000} },
-	{ 0x5dc, { 0x00000000} },
-	{ 0x5e0, { 0x00000000} },
-	{ 0x5e4, { 0x00000000} },
-	{ 0x5e8, { 0x00000000} },
-	{ 0x5ec, { 0x00000000} },
-	{ 0x5f0, { 0x00000000} },
-	{ 0x5f4, { 0x00000000} },
-	{ 0x5f8, { 0x00000000} },
-	{ 0x5fc, { 0x00000000} },
-	{ 0x600, { 0x00000000} },
-	{ 0x604, { 0x00000000} },
-	{ 0x608, { 0x00000000} },
-	{ 0x60c, { 0x00000000} },
-	{ 0x610, { 0x00000000} },
-	{ 0x614, { 0x00000000} },
-	{ 0x618, { 0x00000000} },
-	{ 0x61c, { 0x00000000} },
-	{ 0x620, { 0x00000000} },
-	{ 0x624, { 0x00000000} },
-	{ 0x628, { 0x00000000} },
-	{ 0x62c, { 0x00000000} },
-	{ 0x630, { 0x00000000} },
-	{ 0x634, { 0x00000000} },
-	{ 0x638, { 0x00000000} },
-	{ 0x63c, { 0x00000000} },
-	{ 0x640, { 0x00000000} },
-	{ 0x644, { 0x00000000} },
-	{ 0x648, { 0x00000000} },
-	{ 0x64c, { 0x00000000} },
-	{ 0x650, { 0x03ffffff} },
-	{ 0x654, { 0x00000000} },
-	{ 0x658, { 0x00000000} },
-	{ 0x65c, { 0x00000000} },
-	{ 0x660, { 0x00000000} },
-	{ 0x664, { 0x00000000} },
-	{ 0x668, { 0x00000000} },
-	{ 0x66c, { 0x00000000} },
-	{ 0x670, { 0x00000000} },
-	{ 0x674, { 0x00000000} },
-	{ 0x678, { 0x00000000} },
-	{ 0x67c, { 0x00000000} },
-	{ 0x680, { 0x00000000} },
-	{ 0x684, { 0x00000000} },
-	{ 0x688, { 0x00000000} },
-	{ 0x68c, { 0x00000000} },
-	{ 0x690, { 0x00000000} },
-	{ 0x694, { 0x00000000} },
-	{ 0x698, { 0x00000000} },
-	{ 0x69c, { 0x00000000} },
-	{ 0x6a0, { 0x00000000} },
-	{ 0x6a4, { 0x00000000} },
-	{ 0x6a8, { 0x00000000} },
-	{ 0x6ac, { 0x00000000} },
-	{ 0x6b0, { 0x00000000} },
-	{ 0x6b4, { 0x00000000} },
-	{ 0x6b8, { 0x00000000} },
-	{ 0x6bc, { 0x00000000} },
-	{ 0x6c0, { 0x00000000} },
-	{ 0x6c4, { 0x00000000} },
-	{ 0x6c8, { 0x00000000} },
-	{ 0x6cc, { 0x00000000} },
-	{ 0x6d0, { 0x00000000} },
-	{ 0x6d4, { 0x00000000} },
-	{ 0x6d8, { 0x00000000} },
-	{ 0x6dc, { 0x00000000} },
-	{ 0x6e0, { 0x00000000} },
-	{ 0x6e4, { 0x00000000} },
-	{ 0x6e8, { 0x00000000} },
-	{ 0x6ec, { 0x00000000} },
-	{ 0x6f0, { 0x00000000} },
-	{ 0x6f4, { 0x00000000} },
-	{ 0x6f8, { 0x00000000} },
-	{ 0x6fc, { 0x00000000} },
-	{ 0x700, { 0x00000000} },
-	{ 0x704, { 0x00000000} },
-	{ 0x708, { 0x00000000} },
-	{ 0x70c, { 0x00000000} },
-	{ 0x710, { 0x03000000} },
-	{ 0x714, { 0x00000000} },
-	{ 0x718, { 0x00000000} },
-	{ 0x71c, { 0x00000000} },
-	{ 0x720, { 0x00000000} },
-	{ 0x724, { 0x00000000} },
-	{ 0x728, { 0x00000000} },
-	{ 0x72c, { 0x00000000} },
-	{ 0x730, { 0x00000000} },
-	{ 0x734, { 0x00000000} },
-	{ 0x738, { 0x00000000} },
-	{ 0x73c, { 0x00000000} },
-	{ 0x740, { 0x00000000} },
-	{ 0x744, { 0x00000000} },
-	{ 0x748, { 0x00000000} },
-	{ 0x74c, { 0x00000000} },
-	{ 0x750, { 0x00008000} },
-	{ 0x754, { 0x00000000} },
-	{ 0x758, { 0x00000000} },
-	{ 0x75c, { 0x00000000} },
-	{ 0x760, { 0x00000000} },
-	{ 0x764, { 0x00000000} },
-	{ 0x768, { 0x00000000} },
-	{ 0x76c, { 0x00000000} },
-	{ 0x770, { 0x00000000} },
-	{ 0x774, { 0x00000000} },
-	{ 0x778, { 0x00000000} },
-	{ 0x77c, { 0x00000000} },
-	{ 0x780, { 0x00000000} },
-	{ 0x784, { 0x00000000} },
-	{ 0x788, { 0x00000000} },
-	{ 0x78c, { 0x00000000} },
-	{ 0x790, { 0x00000000} },
-	{ 0x794, { 0x00000000} },
-	{ 0x798, { 0x00000000} },
-	{ 0x79c, { 0x00000000} },
-	{ 0x7a0, { 0x00000000} },
-	{ 0x7a4, { 0x00008010} },
-	{ 0x7a8, { 0x00000000} },
-	{ 0x7ac, { 0x00000000} },
-	{ 0x7b0, { 0x00000000} },
-	{ 0x7b4, { 0x00000000} },
-	{ 0x7b8, { 0x00000000} },
-	{ 0x7bc, { 0x00000000} },
-	{ 0x7c0, { 0x00000000} },
-	{ 0x7c4, { 0x00000000} },
-	{ 0x7c8, { 0x00000000} },
-	{ 0x7cc, { 0x00000000} },
-	{ 0x7d0, { 0x00000000} },
-	{ 0x7d4, { 0x00000000} },
-	{ 0x7d8, { 0x00000000} },
-	{ 0x7dc, { 0x00000000} },
-	{ 0x7e0, { 0x00000000} },
-	{ 0x7e4, { 0x00000000} },
-	{ 0x7e8, { 0x00000000} },
-	{ 0x7ec, { 0x00000000} },
-	{ 0x7f0, { 0x00000000} },
-	{ 0x7f4, { 0x00000000} },
-	{ 0x7f8, { 0x00000000} },
-	{ 0x7fc, { 0x00000000} },
 	{ -1, { -1} }
 };
 
 struct mvebu_mckinley_config mckinley_phy_config[] = {
-/*  reg_offset  650_HZ */
-	{ 0x800, { 0x00044050} },
-	{ 0x804, { 0xe447744b} },
-	{ 0x808, { 0xcff00443} },
-	{ 0x80c, { 0x3703ff44} },
-	{ 0x810, { 0x00300108} },
-	{ 0x814, { 0x00080200} },
-	{ 0x818, { 0x00000000} },
-	{ 0x81c, { 0xa0118000} },
-	{ 0x820, { 0x00000000} },
-	{ 0x824, { 0x00000000} },
-	{ 0x828, { 0x00000000} },
-	{ 0x82c, { 0x00000000} },
-	{ 0x830, { 0x03800000} },
-	{ 0x834, { 0x00000000} },
-	{ 0x838, { 0x5b050003} },
-	{ 0x83c, { 0x00000020} },
-	{ 0x840, { 0x00000400} },
-	{ 0x844, { 0x00000000} },
-	{ 0x848, { 0x00000000} },
-	{ 0x84c, { 0x00000000} },
-	{ 0x850, { 0x00000603} },
-	{ 0x854, { 0x00000603} },
-	{ 0x858, { 0x00000703} },
-	{ 0x85c, { 0x00000603} },
-	{ 0x860, { 0x00000603} },
-	{ 0x864, { 0x00000603} },
-	{ 0x868, { 0x00000603} },
-	{ 0x86c, { 0x00000603} },
-	{ 0x870, { 0x20200000} },
-	{ 0x874, { 0x20200000} },
-	{ 0x878, { 0x00000000} },
-	{ 0x87c, { 0x0060012a} },
-	{ 0x878, { 0x00000001} },
-	{ 0x87c, { 0x00600124} },
-	{ 0x878, { 0x00000002} },
-	{ 0x87c, { 0x0070011a} },
-	{ 0x878, { 0x00000003} },
-	{ 0x87c, { 0x0060011c} },
-	{ 0x878, { 0x00000004} },
-	{ 0x87c, { 0x0060011b} },
-	{ 0x878, { 0x00000005} },
-	{ 0x87c, { 0x00600118} },
-	{ 0x878, { 0x00000006} },
-	{ 0x87c, { 0x0060011a} },
-	{ 0x878, { 0x00000007} },
-	{ 0x87c, { 0x0060011a} },
-	{ 0x878, { 0x00000008} },
-	{ 0x87c, { 0x00900121} },
-	{ 0x878, { 0x00000009} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000a} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000b} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000c} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000d} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000e} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000f} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x890, { 0x0000007b} },
-	{ 0x894, { 0x0000007b} },
-	{ 0x898, { 0x0000007b} },
-	{ 0x89c, { 0x0000007b} },
-	{ 0x8a0, { 0x0000007b} },
-	{ 0x8a4, { 0x0000007b} },
-	{ 0x8a8, { 0x0000007b} },
-	{ 0x8ac, { 0x0000007b} },
-	{ 0x8b0, { 0x0000077b} },
-	{ 0x8b4, { 0x00000000} },
-	{ 0x8b8, { 0x00000000} },
-	{ 0x8bc, { 0x00000000} },
-	{ 0x8c0, { 0x41000000} },
-	{ 0x8e0, { 0x00001001} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x0000003f} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001002} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x0000003f} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001004} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x0000003f} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001008} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x0000002f} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001010} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x00000000} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001020} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x00000000} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001040} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x00000000} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001080} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x00000000} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001100} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x00000000} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001200} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x00000000} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001400} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x0000001e} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001800} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x00000019} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00001000} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x00000019} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x8e0, { 0x00003000} },
-	{ 0xa20, { 0x00000000} },
-	{ 0x8d0, { 0x00000019} },
-	{ 0x8d4, { 0x00000000} },
-	{ 0x8d8, { 0x00000000} },
-	{ 0x8dc, { 0x00000000} },
-	{ 0x980, { 0x00000000} },
-	{ 0x984, { 0x00000000} },
-	{ 0x988, { 0x00000000} },
-	{ 0x98c, { 0x00000000} },
-	{ 0x990, { 0x00000000} },
-	{ 0x994, { 0x00000000} },
-	{ 0x998, { 0x00000000} },
-	{ 0x99c, { 0x00000000} },
-	{ 0x9a0, { 0x00000000} },
-	{ 0x9a4, { 0x00000000} },
-	{ 0x9a8, { 0x00000000} },
-	{ 0x9ac, { 0x00000000} },
-	{ 0x9b0, { 0x00000000} },
-	{ 0x9b4, { 0x00000000} },
-	{ 0x9b8, { 0x00000000} },
-	{ 0x9bc, { 0x00000000} },
-	{ 0x9c0, { 0x00000000} },
-	{ 0x9c4, { 0x00000000} },
-	{ 0x9c8, { 0x00000000} },
-	{ 0x9cc, { 0x00000000} },
-	{ 0x9d0, { 0x00000000} },
-	{ 0x9d4, { 0x00000000} },
-	{ 0x9d8, { 0x00000000} },
-	{ 0x9dc, { 0x00000000} },
-	{ 0x9e0, { 0x00000000} },
-	{ 0x9e4, { 0x00000000} },
-	{ 0x9e8, { 0x00000000} },
-	{ 0x9ec, { 0x00000000} },
-	{ 0x9f0, { 0x00000000} },
-	{ 0x9f4, { 0x00000000} },
-	{ 0x9f8, { 0x00000000} },
-	{ 0x9fc, { 0x00000000} },
-	{ 0xa00, { 0x00000000} },
-	{ 0xa04, { 0x00000000} },
-	{ 0xa08, { 0x00000000} },
-	{ 0xa0c, { 0x00000000} },
-	{ 0xa10, { 0x00000000} },
-	{ 0xa14, { 0x00000000} },
-	{ 0xa18, { 0x00000000} },
-	{ 0xa1c, { 0x00000000} },
-	{ 0xa20, { 0x00000000} },
-	{ 0xa24, { 0x00000000} },
-	{ 0xa28, { 0x00000000} },
-	{ 0xa2c, { 0x00000000} },
-	{ 0xa30, { 0x00000000} },
-	{ 0xa34, { 0x00000000} },
-	{ 0xa38, { 0x00000000} },
-	{ 0xa3c, { 0x00000000} },
-	{ 0xa40, { 0x00000000} },
-	{ 0xa44, { 0x00000000} },
-	{ 0xa48, { 0x00000000} },
-	{ 0xa4c, { 0x00000000} },
-	{ 0xa50, { 0x00000000} },
-	{ 0xa54, { 0x00000000} },
-	{ 0xa58, { 0x00000000} },
-	{ 0xa5c, { 0x00000000} },
-	{ 0xa60, { 0x00000000} },
-	{ 0xa64, { 0x00000000} },
-	{ 0xa68, { 0x00000000} },
-	{ 0xa6c, { 0x00000000} },
-	{ 0xa70, { 0x00000000} },
-	{ 0xa74, { 0x00000000} },
-	{ 0xa78, { 0x00000000} },
-	{ 0xa7c, { 0x00000000} },
-	{ 0xa80, { 0x00000000} },
-	{ 0xa84, { 0x00000000} },
-	{ 0xa88, { 0x00000000} },
-	{ 0xa8c, { 0x00000000} },
-	{ 0xa90, { 0x00000000} },
-	{ 0xa94, { 0x00000000} },
-	{ 0xa98, { 0x00000000} },
-	{ 0xa9c, { 0x00000000} },
-	{ 0xaa0, { 0x00000000} },
-	{ 0xaa4, { 0x00000000} },
-	{ 0xaa8, { 0x00000000} },
-	{ 0xaac, { 0x00000000} },
-	{ 0xab0, { 0x00000000} },
-	{ 0xab4, { 0x00000000} },
-	{ 0xab8, { 0x00000000} },
-	{ 0xabc, { 0x00000000} },
-	{ 0xac0, { 0x00000000} },
-	{ 0xac4, { 0x00000000} },
-	{ 0xac8, { 0x00000000} },
-	{ 0xacc, { 0x00000000} },
-	{ 0xad0, { 0x00000000} },
-	{ 0xad4, { 0x00000000} },
-	{ 0xad8, { 0x00000000} },
-	{ 0xadc, { 0x00000000} },
-	{ 0xae0, { 0x00000000} },
-	{ 0xae4, { 0x00000000} },
-	{ 0xae8, { 0x00000000} },
-	{ 0xaec, { 0x00000000} },
-	{ 0xaf0, { 0x00000000} },
-	{ 0xaf4, { 0x00000000} },
-	{ 0xaf8, { 0x00000000} },
-	{ 0xafc, { 0x00000000} },
-	{ 0xb00, { 0x00000000} },
-	{ 0xb04, { 0x00000000} },
-	{ 0xb08, { 0x00000000} },
-	{ 0xb0c, { 0x00000000} },
-	{ 0xb10, { 0x00000000} },
-	{ 0xb14, { 0x00000000} },
-	{ 0xb18, { 0x00000000} },
-	{ 0xb1c, { 0x00000000} },
-	{ 0xb20, { 0x00000000} },
-	{ 0xb24, { 0x00000000} },
-	{ 0xb28, { 0x00000000} },
-	{ 0xb2c, { 0x00000000} },
-	{ 0xb30, { 0x00000000} },
-	{ 0xb34, { 0x00000000} },
-	{ 0xb38, { 0x00000000} },
-	{ 0xb3c, { 0x00000000} },
-	{ 0xb40, { 0x00000000} },
-	{ 0xb44, { 0x00000000} },
-	{ 0xb48, { 0x00000000} },
-	{ 0xb4c, { 0x00000000} },
-	{ 0xb50, { 0x00000000} },
-	{ 0xb54, { 0x00000000} },
-	{ 0xb58, { 0x00000000} },
-	{ 0xb5c, { 0x00000000} },
-	{ 0xb60, { 0x00000000} },
-	{ 0xb64, { 0x00000000} },
-	{ 0xb68, { 0x00000000} },
-	{ 0xb6c, { 0x00000000} },
-	{ 0xb70, { 0x00000000} },
-	{ 0xb74, { 0x00000000} },
-	{ 0xb78, { 0x00000000} },
-	{ 0xb7c, { 0x00000000} },
-	{ 0xb80, { 0x00000000} },
-	{ 0xb84, { 0x00000000} },
-	{ 0xb88, { 0x00000000} },
-	{ 0xb8c, { 0x00000000} },
-	{ 0xb90, { 0x00000000} },
-	{ 0xb94, { 0x00000000} },
-	{ 0xb98, { 0x00000000} },
-	{ 0xb9c, { 0x00000000} },
-	{ 0xba0, { 0x02300000} },
-	{ 0xba4, { 0x0c6318c6} },
-	{ 0x804, { 0x10777ffa} },
-	{ 0x808, { 0x0ff00ff2} },
-	{ 0x80c, { 0x4048ccff} },
-	{ 0x828, { 0x00000000} },
-	{ 0x830, { 0x03800000} },
-	{ 0x834, { 0x00000000} },
-	{ 0x810, { 0x00100100} },
-	{ 0x840, { 0x00000400} },
-	{ 0x890, { 0x00000000} },
-	{ 0x894, { 0x00000000} },
-	{ 0x898, { 0x00000000} },
-	{ 0x89c, { 0x00000000} },
-	{ 0x8a0, { 0x00000000} },
-	{ 0x8a4, { 0x00000000} },
-	{ 0x8a8, { 0x00000000} },
-	{ 0x8ac, { 0x00000000} },
-	{ 0x800, { 0x000440f0} },
-	{ 0x814, { 0x00080200} },
-	{ 0x838, { 0x00000003} },
-	{ 0x83c, { 0x00000020} },
-	{ 0x8e0, { 0x00010001} },
-	{ 0x8d0, { 0x0000003f} },
-	{ 0x8e0, { 0x00010002} },
-	{ 0x8d0, { 0x0000003f} },
-	{ 0x8e0, { 0x00010004} },
-	{ 0x8d0, { 0x0000003f} },
-	{ 0x8e0, { 0x00010008} },
-	{ 0x8d0, { 0x0000002f} },
-	{ 0x8e0, { 0x00010400} },
-	{ 0x8d0, { 0x0000001e} },
-	{ 0x8e0, { 0x00010800} },
-	{ 0x8d0, { 0x00000019} },
-	{ 0x8e0, { 0x00011000} },
-	{ 0x8d0, { 0x00000019} },
-	{ 0x8e0, { 0x00012000} },
-	{ 0x8d0, { 0x00000019} },
-	{ 0x850, { 0x00000b03} },
-	{ 0x854, { 0x00000b03} },
-	{ 0x858, { 0x00000b03} },
-	{ 0x85c, { 0x00000b03} },
-	{ 0x860, { 0x00000b03} },
-	{ 0x864, { 0x00000b03} },
-	{ 0x868, { 0x00000b03} },
-	{ 0x86c, { 0x00000b03} },
-	{ 0x820, { 0x20000000} },
-	{ 0x820, { 0x40000000} },
-	{ 0x838, { 0x1c050003} },
-	{ 0x8c0, { 0xc1000000} },
-	{ 0x98 , { 0x0040004e} },
-	{ 0x108, { 0xffff0002} },
-	{ 0x204, { 0x00000000} },
-	{ 0x220, { 0x11010642} },
-	{ 0x2c0, { 0x0000e000} },
-	{ 0x2c4, { 0x00000020} },
-	{ 0x300, { 0x00000709} },
-	{ 0x304, { 0x00000000} },
-	{ 0x310, { 0x00110000} },
-	{ 0x340, { 0x00000001} },
-	{ 0x344, { 0x00000002} },
-	{ 0x380, { 0x000516bc} },
-	{ 0x384, { 0x00020918} },
-	{ 0x388, { 0x00000043} },
-	{ 0x38c, { 0x00000200} },
-	{ 0x390, { 0x00400100} },
-	{ 0x394, { 0x00ea0517} },
-	{ 0x398, { 0x00f10200} },
-	{ 0x39c, { 0x00000707} },
-	{ 0x3a0, { 0x00050511} },
-	{ 0x3a4, { 0x00000002} },
-	{ 0x3a8, { 0x00000c04} },
-	{ 0x3ac, { 0x18200a17} },
-	{ 0x3b0, { 0x0a0b060a} },
-	{ 0x3b4, { 0x04000600} },
-	{ 0x3b8, { 0x00000500} },
-	{ 0x3bc, { 0x00050000} },
-	{ 0x3c0, { 0x00000000} },
-	{ 0x3c4, { 0x00000000} },
-	{ 0x3c8, { 0x00000000} },
-	{ 0x3cc, { 0x00000000} },
-	{ 0x3d0, { 0x00000000} },
-	{ 0x800, { 0x00044050} },
-	{ 0x804, { 0xe447744b} },
-	{ 0x808, { 0xcff00443} },
-	{ 0x80c, { 0x3703ff44} },
-	{ 0x810, { 0x00300108} },
-	{ 0x814, { 0x00080200} },
-	{ 0x818, { 0x00000000} },
-	{ 0x81c, { 0xa0118000} },
-	{ 0x820, { 0x00a18004} },
-	{ 0x824, { 0x000001ff} },
-	{ 0x828, { 0x00000000} },
-	{ 0x82c, { 0x00000000} },
-	{ 0x830, { 0x03800000} },
-	{ 0x834, { 0x00000000} },
-	{ 0x838, { 0x60040043} },
-	{ 0x83c, { 0x00000020} },
-	{ 0x840, { 0x00000400} },
-	{ 0x844, { 0x00000000} },
-	{ 0x848, { 0x00000000} },
-	{ 0x84c, { 0x00000000} },
-	{ 0x850, { 0x20200000} },
-	{ 0x854, { 0x20200000} },
-	{ 0x858, { 0x20200000} },
-	{ 0x85c, { 0x20200000} },
-	{ 0x860, { 0x20200000} },
-	{ 0x864, { 0x20200000} },
-	{ 0x868, { 0x20200000} },
-	{ 0x86c, { 0x20200000} },
-	{ 0x870, { 0x20200000} },
-	{ 0x874, { 0x20200000} },
-	{ 0x878, { 0x00000000} },
-	{ 0x87c, { 0x0095012b} },
-	{ 0x878, { 0x00000001} },
-	{ 0x87c, { 0x0095012d} },
-	{ 0x878, { 0x00000002} },
-	{ 0x87c, { 0x00910122} },
-	{ 0x878, { 0x00000003} },
-	{ 0x87c, { 0x008f011e} },
-	{ 0x878, { 0x00000004} },
-	{ 0x87c, { 0x00910121} },
-	{ 0x878, { 0x00000005} },
-	{ 0x87c, { 0x008f011a} },
-	{ 0x878, { 0x00000006} },
-	{ 0x87c, { 0x008f011e} },
-	{ 0x878, { 0x00000007} },
-	{ 0x87c, { 0x00900122} },
-	{ 0x878, { 0x00000008} },
-	{ 0x87c, { 0x00900123} },
-	{ 0x878, { 0x00000009} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000a} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000b} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000c} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000d} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000e} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x878, { 0x0000000f} },
-	{ 0x87c, { 0x00000000} },
-	{ 0x890, { 0x0000007b} },
-	{ 0x894, { 0x0000007b} },
-	{ 0x898, { 0x0000007b} },
-	{ 0x89c, { 0x0000007b} },
-	{ 0x8a0, { 0x0000007b} },
-	{ 0x8a4, { 0x0000007b} },
-	{ 0x8a8, { 0x0000007b} },
-	{ 0x8ac, { 0x0000007b} },
-	{ 0x8b0, { 0x0000077b} },
-	{ 0x8b4, { 0x00000000} },
-	{ 0x8b8, { 0x00000000} },
-	{ 0x8bc, { 0x00000000} },
-	{ 0x8c0, { 0x41000000} },
-	{ 0x820, { 0x20000000} },
-	{ 0x820, { 0x40000000} },
-	{ 0x820, { 0x80000000} },
-	{ 0x850, { 0x00000603} },
-	{ 0x854, { 0x00000603} },
-	{ 0x858, { 0x00000703} },
-	{ 0x85c, { 0x00000603} },
-	{ 0x860, { 0x00000603} },
-	{ 0x864, { 0x00000603} },
-	{ 0x868, { 0x00000603} },
-	{ 0x86c, { 0x00000603} },
-	{ 0x838, { 0x5b050003} },
-	{ 0x8e0, { 0x00010001} },
-	{ 0x8d0, { 0x0000003f} },
-	{ 0x8e0, { 0x00010002} },
-	{ 0x8d0, { 0x0000003f} },
-	{ 0x8e0, { 0x00010004} },
-	{ 0x8d0, { 0x0000003f} },
-	{ 0x8e0, { 0x00010008} },
-	{ 0x8d0, { 0x0000002f} },
-	{ 0x8e0, { 0x00010400} },
-	{ 0x8d0, { 0x0000001e} },
-	{ 0x8e0, { 0x00010800} },
-	{ 0x8d0, { 0x00000019} },
-	{ 0x8e0, { 0x00011000} },
-	{ 0x8d0, { 0x00000019} },
-	{ 0x8e0, { 0x00012000} },
-	{ 0x8d0, { 0x00000019} },
 	{-1, {-1} },
 };
 
@@ -1173,3 +188,178 @@ void mvebu_dram_phy_init(struct mvebu_dram_config *dram_config)
 #endif
 	debug_exit();
 }
+
+
+static void mvebu_dram_dll_set(unsigned short dll_phsel, unsigned short dll_phsel1, void *base_addr)
+{
+	debug_enter();
+
+	debug("set dll_phsel=%#x, dll_phsel1=%#x\n", dll_phsel, dll_phsel1);
+
+	mck6_writel((dll_phsel << 16) | (dll_phsel1 << 24), base_addr + MC6_CH0_PHY_DLL_CONTROL_B0 + 0);
+	mck6_writel((dll_phsel << 16) | (dll_phsel1 << 24), base_addr + MC6_CH0_PHY_DLL_CONTROL_B0 + 4);
+	mck6_writel((dll_phsel << 16) | (dll_phsel1 << 24), base_addr + MC6_CH0_PHY_DLL_CONTROL_B0 + 8);
+	mck6_writel((dll_phsel << 16) | (dll_phsel1 << 24), base_addr + MC6_CH0_PHY_DLL_CONTROL_B0 + 12);
+
+	/* Updates DLL master. Block read/MMR for 4096 MCLK cycles to guarantee DLL lock.
+	   Either wait 4096 MCLK (memPll/4) cycles, or check DLL lock status
+	 */
+	mck6_writel(0x20000000, base_addr + MC6_CH0_PHY_CONTROL_9);
+
+	/* Reset PHY DLL. Dll_reset_timer ([31:24] of PHY Control Register 8,
+	   Offset 0x41C/0xC1C) is set to 0x10, reset DLL for 128*32=4096 MCLK cycles.
+	 */
+	udelay(100);
+
+	/* Copy DLL master to DLL slave. Slave controls the actual delay_l.
+	   Both DLL bypass and DLL needs 'update'.
+	 */
+	mck6_writel(0x40000000, base_addr + MC6_CH0_PHY_CONTROL_9);
+
+	/* Update Dll delay_l. When Dll_update_timer ([20:16] of PHY Control Register 8,
+	   Offset 0x41C/0xC1C) is 0x11, assert DLL_UPDATE_EN pin for 0x11*16 = 272 MCLK cycles.
+	 */
+	udelay(50);
+
+	/* KW Finish DLL update*/
+	mck6_writel(0x0, base_addr + MC6_CH0_PHY_CONTROL_9);
+
+	debug_exit();
+}
+
+static unsigned int mvebu_dram_dll_wr_test(u64 start, u64 size)
+{
+	u64 i, t;
+	u64 *ptr = NULL;
+
+	for (i = 0; i < size; i += sizeof(u64))
+		*((u64 *)((u64)(start + i))) = start + i;
+
+	for (i = 0; i < size; i += sizeof(u64)) {
+		ptr = (u64 *)((u64)(start + i));
+		t = *(ptr);
+		if (t != start + i) {
+			debug("Error: [%08llX] expect %08llX, get %08llX\n", start + i, start + i, t);
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static unsigned short mvebu_dram_dll_search(unsigned short dll, unsigned short round, void *base_addr)
+{
+	unsigned short dll_var, dll_phsel, dll_phsel1;
+	unsigned int   optimal_rd_dll;
+	unsigned int   MIN_RD_DLL = 0xffff, MAX_RD_DLL = 0xffff;
+
+	debug_enter();
+
+	for (dll_var = DLL_PHSEL_START; dll_var < DLL_PHSEL_END; dll_var += DLL_PHSEL_STEP) {
+		if (round == 0) {
+			dll_phsel = dll_var;
+			dll_phsel1 = dll_var;
+		} else if (round == 1) {
+			dll_phsel = dll;	/*Pos*/
+			dll_phsel1 = dll_var;
+		} else {
+			dll_phsel = dll_var;
+			dll_phsel1 = dll;	/*Neg*/
+		}
+		/* Set dll */
+		mvebu_dram_dll_set(dll_phsel, dll_phsel1, base_addr);
+
+		if (!mvebu_dram_dll_wr_test(SDRAM_DIRECT_START, 1024 * 2)) {/*pass test*/
+			if (MIN_RD_DLL == 0xffff)
+				MIN_RD_DLL = dll_var;
+			MAX_RD_DLL = dll_var;
+			debug("Search RD_DLL Pass: DDR CTL = %#x.\n", dll_var);
+		} else {
+			debug("Search RD_DLL fail: DDR CTL = %#x.\n", dll_var);
+			if (MIN_RD_DLL != 0xffff)
+				break;
+		}
+		debug(" RD_DLL = 0x%x.\n", dll_var);
+		if (dll_var == 0x3f)
+			break;
+	} /* end of phase loop */
+
+	if (MIN_RD_DLL == 0xFFFF) {
+		debug("DDR: No DLL found.\n");
+		optimal_rd_dll = 0xffff;
+	} else {
+		optimal_rd_dll =  (MAX_RD_DLL - MIN_RD_DLL)/2 + MIN_RD_DLL;
+		debug("DDR: end DLL tuning - MIN = %#x, MAX = %#x, optimal = %#x\n",
+			MIN_RD_DLL, MAX_RD_DLL, optimal_rd_dll);
+		if (round == 0) {
+			dll_phsel = optimal_rd_dll;
+			dll_phsel1 = optimal_rd_dll;
+		} else if (round == 1) {
+			dll_phsel = dll;		/*Pos*/
+			dll_phsel1 = optimal_rd_dll;
+		} else {
+			dll_phsel = optimal_rd_dll;
+			dll_phsel1 = dll;		/*Neg*/
+		}
+		/* Set dll */
+		mvebu_dram_dll_set(dll_phsel - 5, dll_phsel1 - 5, base_addr);
+	}
+
+	debug_exit();
+
+	return optimal_rd_dll;
+}
+
+void mvebu_dram_dll_tune(struct mvebu_dram_config *dram_config)
+{
+	unsigned short dll_median, dll_phsel, dll_phsel1;
+	unsigned int regval, dll, status = 0;
+	void __iomem *base_addr = dram_config->phy_base;
+
+	debug_enter();
+
+	dll = readl(base_addr + MC6_CH0_PHY_DLL_CONTROL_B0);
+	printf("DDR: start DLL tuning with initial phase delays (P) %#x, (N) %#x\n",
+		(dll >> 16) & 0x3F, (dll >> 24) & 0x3F);
+
+	dll = (dll >> 16) & 0x3F;
+	mck6_writel(0x0, base_addr + MC6_CH0_PHY_CONTROL_9);
+
+	/* Automatically update PHY DLL with interval time set in Dll_auto_update_interval
+	   ([15:8] of PHY Control Register 13, Offset 0x248)
+	 */
+	regval = readl(base_addr + MC6_CH0_PHY_CONTROL_8);
+	/* Turn off Dll_auto_manual_update & Dll_auto_update_en
+	   DLL_auto_update_en has a known bug. Don't use.
+	 */
+	regval &= ~0xC;
+	/* change Dll_reset_timer to 128*32 cycles*/
+	regval |= 0x80000000;
+	mck6_writel(regval, base_addr + MC6_CH0_PHY_CONTROL_8);
+	dll_median = mvebu_dram_dll_search(0, 0, base_addr);
+
+	if (dll_median == 0xffff) {
+		status = 1;
+	} else {/* Find Neg dll */
+		dll_phsel1 = mvebu_dram_dll_search(dll_median, 1, base_addr);
+
+		if (dll_phsel1 == 0xffff) {
+			status = 1;
+		} else {/* Find Pos dll */
+			dll_phsel = mvebu_dram_dll_search(dll_phsel1, 2, base_addr);
+
+			if (dll_phsel == 0xffff)
+				status = 1;
+		}
+	}
+
+	if (status) {
+		printf("DDR: using initial DLL phase delays\n");
+		mvebu_dram_dll_set(dll, dll, base_addr);
+	} else
+		printf("DDR: set new DLL phase delays (P) %#x, (N) %#x\n", dll_phsel, dll_phsel1);
+
+
+	debug_exit();
+
+	return;
+}
diff --git a/drivers/ddr/mvebu_ddr.c b/drivers/ddr/mvebu_ddr.c
index 8f5a648..bf6b6f7 100644
--- a/drivers/ddr/mvebu_ddr.c
+++ b/drivers/ddr/mvebu_ddr.c
@@ -30,9 +30,12 @@
 #ifdef CONFIG_MVEBU_STATIC_SUPPORT
 static void static_dram_init(struct mvebu_dram_config *config)
 {
+#ifndef CONFIG_TARGET_ARMADA_LP
 	mvebu_dram_mac_init(config);
 	mvebu_dram_phy_init(config);
-
+#else
+	mvebu_dram_dll_tune(config);
+#endif
 	printf("Static DRAM initialization is DONE..\n");
 }
 #endif
-- 
1.9.1

